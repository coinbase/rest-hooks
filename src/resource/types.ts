import { schemas, Schema, SchemaMany, SchemaOne } from './normal';
import { RequestOptions } from '~/types';

/** Defines the shape of a network request */
export interface FetchShape<
  S extends Schema,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object> | void = Readonly<object> | undefined
> {
  readonly type: 'read' | 'mutate' | 'delete';
  fetch(params: Params, body: Body): Promise<any>;
  getFetchKey(params: Params): string;
  readonly schema: S;
  readonly options?: RequestOptions;
}

/** Purges a value from the server */
export interface DeleteShape<
  S extends schemas.Entity,
  Params extends Readonly<object> = Readonly<object>
> extends FetchShape<S, Params, any> {
  readonly type: 'delete';
  readonly schema: S;
}

/** To change values on the server */
export interface MutateShape<
  S extends Schema,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object> | void = Readonly<object> | undefined
> extends FetchShape<S, Params, Body> {
  readonly type: 'mutate';
}

/** For retrieval requests */
export interface ReadShape<
  S extends Schema,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object> | void = Readonly<object> | undefined
> extends FetchShape<S, Params, Body> {
  readonly type: 'read';
}

export function isDeleteShape(
  shape: FetchShape<any, any, any>,
): shape is DeleteShape<any, any> {
  return shape.type === 'delete';
}

export type ResultShape<RS> = RS extends { schema: infer U } ? U : never;
export type SelectReturn<RS> = RS extends {
  select: (...args: any[]) => infer U;
}
  ? U
  : never;
export type AlwaysSelect<RS> = NonNullable<SelectReturn<RS>>;
export type ParamArg<RS> = RS extends {
  getFetchKey: (params: infer U) => any;
}
  ? U
  : never;
export type BodyArg<RS> = RS extends {
  fetch: (url: any, body: infer U) => any;
}
  ? U
  : never;
export type RequestResource<RS> = SchemaOf<ResultShape<RS>>;

export function isEntity<T>(schema: Schema): schema is schemas.Entity<T> {
  return (schema as schemas.Entity<T>).key !== undefined;
}

export type SchemaOf<T> = T extends SchemaMany<infer R>
  ? R[]
  : T extends SchemaOne<infer R>
  ? R
  : never;
