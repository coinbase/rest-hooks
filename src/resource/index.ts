import Resource from './Resource';
import {
  SchemaOf,
  DeleteShape,
  ReadShape,
  MutateShape,
  FetchShape,
  isDeleteShape,
  RequestResource,
  isEntity,
} from './types';
import {
  SchemaArray,
  Schema,
  SchemaBase,
  normalize,
  denormalize,
  schemas,
} from './normal';

export type DeleteShape<
  S extends schemas.Entity,
  Params extends Readonly<object> = Readonly<object>
> = DeleteShape<S, Params>;
export type MutateShape<
  S extends Schema,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object> | void = Readonly<object> | undefined
> = MutateShape<S, Params, Body>;
export type ReadShape<
  S extends Schema,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object> | void = Readonly<object> | undefined
> = ReadShape<S, Params, Body>;
export type FetchShape<
  S extends Schema,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object> | void = Readonly<object> | undefined
> = FetchShape<S, Params, Body>;
export type Schema<T = any> = Schema<T>;
export type SchemaOf<T> = SchemaOf<T>;
export type SchemaArray<T> = SchemaArray<T>;
export type SchemaBase<T> = SchemaBase<T>;
export type RequestResource<RS> = RequestResource<RS>;

export { Resource, isEntity, normalize, denormalize, isDeleteShape, schemas };
