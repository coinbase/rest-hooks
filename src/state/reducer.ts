import { normalize } from '../resource';
import { mergeWith } from 'lodash';
import { ActionTypes, State } from '../types';

export const initialState: State<unknown> = {
  entities: {},
  results: {},
  meta: {},
};

type Writable<T> = { [P in keyof T]: NonNullable<T[P]> };

interface MergeableStatic<T> {
  new (): T;
  merge(a: T, b: T): T;
}

function isMergeable<T>(constructor: any): constructor is MergeableStatic<T> {
  return constructor && typeof constructor.merge === 'function';
}

export const resourceCustomizer = (a: any, b: any): any => {
  const Static = b && b.constructor;
  if (a && Static && isMergeable(Static)) {
    return Static.merge(a, b);
  }

  // use default merging in lodash.merge()
};

export default function reducer(state: State<unknown> | undefined, action: ActionTypes) {
  if (!state) state = initialState;
  switch (action.type) {
  case 'rest-hooks/receive':
    if (action.error) {
      return {
        ...state,
        meta: {
          ...state.meta,
          [action.meta.url]: {
            date: action.meta.date,
            error: action.payload,
            expiresAt: action.meta.expiresAt,
          },
        },
      };
    }
    const normalized = normalize(action.payload, action.meta.schema);
    return {
      entities: mergeWith(
        { ...state.entities },
        normalized.entities,
        resourceCustomizer,
      ),
      results: {
        ...state.results,
        [action.meta.url]: normalized.result,
      },
      meta: {
        ...state.meta,
        [action.meta.url]: {
          date: action.meta.date,
          expiresAt: action.meta.expiresAt,
        },
      },
    };
  case 'rest-hooks/rpc':
    if (action.error) return state;
    let { entities } = normalize(action.payload, action.meta.schema);
    return {
      ...state,
      entities: mergeWith(
        { ...state.entities },
        entities,
        resourceCustomizer,
      ),
    };
  case 'rest-hooks/purge':
    if (action.error) return state;
    const key = action.meta.schema.key;
    const pk = action.meta.url;
    const e: Writable<typeof state.entities> = {};
    for (const k in state.entities) {
      const o = state.entities[k];
      if (o === undefined) continue;
      if (k === key) {
        e[k] = {};
        for (const j in o) {
          if (j === pk) continue;
          e[k][j] = o[j];
        }
      } else {
        e[k] = o;
      }
    }
    return {
      ...state,
      entities: e,
    };
  default:
    // A reducer must always return a valid state.
    // Alternatively you can throw an error if an invalid action is dispatched.
    return state;
  }
}
