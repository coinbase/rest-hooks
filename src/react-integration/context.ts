import React from 'react';
import { ActionTypes } from 'types';
import { initialState } from '~/state/reducer';

export const StateContext = React.createContext(initialState);
export const DispatchContext = React.createContext((value: ActionTypes) => {});
