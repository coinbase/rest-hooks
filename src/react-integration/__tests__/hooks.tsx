import React, { Suspense, useEffect } from 'react';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import nock from 'nock';
import { normalize } from '../../resource';

import { DispatchContext, StateContext } from '../context';
import {
  CoolerArticleResource,
  UserResource,
  PaginatedArticleResource,
  ArticleResourceWithOtherListUrl,
  StaticArticleResource,
  InvalidIfStaleArticleResource,
} from '../../__tests__/common';
import {
  useFetcher,
  useRetrieve,
  useResource,
  useCache,
  useResultCache,
  useInvalidator,
  useResetter,
} from '../hooks';
import { initialState } from '../../state/reducer';
import { State, ActionTypes } from '../../types';
import makeRenderRestHook from '../../test/makeRenderRestHook';
import { makeCacheProvider } from '../../test/providers';
import mockInitialState from '../../test/mockState';
import { users, articlesPages, payload } from './fixtures';

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const tree = (
    <DispatchContext.Provider value={dispatch}>
      <Suspense fallback={null}>
        <Component />
      </Suspense>
    </DispatchContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalledTimes(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function testRestHook(
  callback: () => void,
  state: State<unknown>,
  dispatch = (v: ActionTypes) => {},
) {
  return renderHook(callback, {
    wrapper: function Wrapper({ children }) {
      return (
        <DispatchContext.Provider value={dispatch}>
          <StateContext.Provider value={state}>
            {children}
          </StateContext.Provider>
        </DispatchContext.Provider>
      );
    },
  });
}

function ArticleComponentTester({ invalidIfStale = false }) {
  const resource = invalidIfStale
    ? InvalidIfStaleArticleResource
    : CoolerArticleResource;
  const article = useResource(resource.detailShape(), {
    id: payload.id,
  });
  return (
    <div>
      <h3>{article.title}</h3>
      <p>{article.content}</p>
    </div>
  );
}

describe('useFetcher', () => {
  const payload = { id: 1, content: 'hi' };

  it('should dispatch an action that fetches a create', async () => {
    nock('http://test.com')
      .post(`/article-cooler/`)
      .reply(201, payload);

    function DispatchTester() {
      const a = useFetcher(CoolerArticleResource.createShape());
      a({ content: 'hi' }, {});
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should dispatch an action with updater in the meta if update shapes params are passed in', async () => {
    nock('http://test.com')
      .post(`/article-cooler/`)
      .reply(201, payload);

    function DispatchTester() {
      const create = useFetcher(CoolerArticleResource.createShape());
      const params = { content: 'hi' };
      create(params, {}, [
        [
          CoolerArticleResource.listShape(),
          {},
          (article: any, articles: any) => [...articles, article],
        ],
      ]);
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should dispatch an action with multiple updaters in the meta if update shapes params are passed in', async () => {
    nock('http://test.com')
      .post(`/article/`)
      .reply(201, payload);

    function DispatchTester() {
      const create = useFetcher(ArticleResourceWithOtherListUrl.createShape());
      const params = { content: 'hi' };
      create(params, {}, [
        [
          ArticleResourceWithOtherListUrl.listShape(),
          {},
          (article: any, articles: any) => [...articles, article],
        ],
        [
          ArticleResourceWithOtherListUrl.otherListShape(),
          {},
          (article: any, articles: any) => [...articles, article],
        ],
      ]);
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should console.error() a warning when fetching without a Provider', () => {
    const oldError = console.error;
    const spy = (console.error = jest.fn());
    renderHook(() => {
      const a = useFetcher(CoolerArticleResource.createShape());
      a({ content: 'hi' }, {});
      return null;
    });
    expect(spy.mock.calls[0]).toMatchInlineSnapshot(`
      Array [
        "It appears you are trying to use Rest Hooks without a provider.
      Follow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component",
      ]
    `);
    console.error = oldError;
  });

  it('should dispatch an action that fetches a partial update', async () => {
    nock('http://test.com')
      .patch(`/article-cooler/1`)
      .reply(200, payload);

    function DispatchTester() {
      const a = useFetcher(CoolerArticleResource.partialUpdateShape());
      a({ content: 'changed' }, { id: payload.id });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should dispatch an action that fetches a full update', async () => {
    nock('http://test.com')
      .put(`/article-cooler/1`)
      .reply(200, payload);

    function DispatchTester() {
      const a = useFetcher(CoolerArticleResource.updateShape());
      a({ content: 'changed' }, { id: payload.id });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });
});

describe('useInvalidate', () => {
  it('should not invalidate anything if params is null', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let invalidate: any;
    testRestHook(
      () => {
        invalidate = useInvalidator(PaginatedArticleResource.listShape());
      },
      state,
      dispatch,
    );
    invalidate(null);
    expect(dispatch).not.toHaveBeenCalled();
  });
  it('should return a function that dispatches an action to invalidate a resource', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let invalidate: any;
    testRestHook(
      () => {
        invalidate = useInvalidator(PaginatedArticleResource.listShape());
      },
      state,
      dispatch,
    );
    invalidate({});
    expect(dispatch).toHaveBeenCalledWith({
      type: 'rest-hooks/invalidate',
      meta: {
        url: 'GET http://test.com/article-paginated/',
      },
    });
  });
  it('should return the same === function each time', () => {
    const track = jest.fn();

    const { rerender } = renderHook(() => {
      const invalidate = useInvalidator(PaginatedArticleResource.listShape());
      useEffect(track, [invalidate]);
    });
    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 4; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});

describe('useResetter', () => {
  it('should return a function that dispatches an action to reset the cache', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let reset: any;
    testRestHook(
      () => {
        reset = useResetter();
      },
      state,
      dispatch,
    );
    reset({});
    expect(dispatch).toHaveBeenCalledWith({
      type: 'rest-hooks/reset',
    });
  });
  it('should return the same === function each time', () => {
    const track = jest.fn();

    const { rerender } = renderHook(() => {
      const reset = useResetter();
      useEffect(track, [reset]);
    });
    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 4; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});

describe('useCache', () => {
  it('should select singles', () => {
    let article: any;
    let state = { ...initialState };
    const { rerender } = renderHook(
      () => (article = useCache(CoolerArticleResource.detailShape(), payload)),
      {
        wrapper: function Wrapper({ children }) {
          return (
            <StateContext.Provider value={state}>
              {children}
            </StateContext.Provider>
          );
        },
      },
    );
    expect(article).toBe(null);
    state = mockInitialState([
      {
        request: CoolerArticleResource.detailShape(),
        params: payload,
        result: payload,
      },
    ]);
    rerender();
    expect(article).toBeTruthy();
    expect(article.title).toBe(payload.title);
  });

  it('should select paginated results', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    let articles: any;
    testRestHook(() => {
      articles = useCache(PaginatedArticleResource.listShape(), {});
    }, state);
    expect(articles).toBeDefined();
    expect(articles.length).toBe(articlesPages.results.length);
    expect(articles[0]).toBeInstanceOf(PaginatedArticleResource);
    expect(articles).toMatchSnapshot();
  });

  it('should return identical value no matter how many re-renders', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const track = jest.fn();

    const { rerender } = testRestHook(() => {
      const articles = useCache(PaginatedArticleResource.listShape(), {});
      useEffect(track, [articles]);
    }, state);

    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 2; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});

describe('useResultCache', () => {
  it('should be null with nothing in state', () => {
    let results: any;
    const state = { ...initialState };
    const { rerender } = testRestHook(() => {
      results = useResultCache(PaginatedArticleResource.listShape(), {});
    }, state);
    expect(results).toBe(null);
  });

  it('should send defaults with nothing in state', () => {
    let results: any;
    const state = { ...initialState };
    const defaults = { prevPage: '', nextPage: '' };
    testRestHook(() => {
      results = useResultCache(
        PaginatedArticleResource.listShape(),
        {},
        defaults,
      );
    }, state);
    expect(results).toEqual(defaults);
  });

  it('should find results', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    let results: any;
    testRestHook(() => {
      results = useResultCache(PaginatedArticleResource.listShape(), {});
    }, state);
    expect(results).toBeTruthy();
    expect(results.nextPage).toBe(articlesPages.nextPage);
    expect(results.prevPage).toBe(articlesPages.prevPage);
    expect(results.results).toEqual(['23', '44', '2', '643']);
  });

  it('should return identical value no matter how many re-renders', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const track = jest.fn();

    const { rerender } = testRestHook(() => {
      const results = useResultCache(PaginatedArticleResource.listShape(), {});
      useEffect(track, [results]);
    }, state);

    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 2; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});

describe('useRetrieve', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  beforeEach(() => {
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload);
    nock('http://test.com')
      .get(`/article-static/${payload.id}`)
      .reply(200, payload);
    nock('http://test.com')
      .get(`/user/`)
      .reply(200, users);
    renderRestHook = makeRenderRestHook(makeCacheProvider);
  });
  afterEach(() => {
    nock.cleanAll();
    renderRestHook.cleanup();
  });

  it('should dispatch singles', async () => {
    function FetchTester() {
      useRetrieve(CoolerArticleResource.detailShape(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not dispatch will null params', () => {
    const dispatch = jest.fn();
    let params: any = null;
    const { rerender } = testRestHook(
      () => {
        useRetrieve(CoolerArticleResource.detailShape(), params);
      },
      initialState,
      dispatch,
    );
    expect(dispatch).toBeCalledTimes(0);
    params = payload;
    rerender();
    expect(dispatch).toBeCalled();
  });

  it('should dispatch with resource defined dataExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.detailShape(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined dataExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.longLivingRequest(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined errorExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.neverRetryOnErrorRequest(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not refetch after expiry and render', async () => {
    let time = 1000;
    global.Date.now = jest.fn(() => time);
    nock.cleanAll();
    const fetchMock = jest.fn(() => payload);
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .reply(200, fetchMock)
      .persist();
    const results: any[] = [
      {
        request: CoolerArticleResource.detailShape(),
        params: payload,
        result: payload,
      },
    ];
    const { result, rerender } = renderRestHook(
      () => {
        return useRetrieve(CoolerArticleResource.detailShape(), payload);
      },
      { results },
    );
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    time += 100;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    // eslint-disable-next-line require-atomic-updates
    time += 610000000;
    rerender();
    await result.current;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
  });
});

describe('useResource()', () => {
  const fbmock = jest.fn();

  function Fallback() {
    fbmock();
    return null;
  }

  beforeEach(() => {
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload);
    nock('http://test.com')
      .get(`/user/`)
      .reply(200, users);
  });

  it('should dispatch an action that fetches', async () => {
    await testDispatchFetch(ArticleComponentTester, [payload]);
  });

  it('should dispatch fetch when sent multiple arguments', async () => {
    function MultiResourceTester() {
      const [article, user] = useResource(
        [
          CoolerArticleResource.detailShape(),
          {
            id: payload.id,
          },
        ],
        [UserResource.listShape(), {}],
      );
      return null;
    }
    await testDispatchFetch(MultiResourceTester, [payload, users]);
  });
  it('should throw same promise until both resolve', async () => {
    const renderRestHook = makeRenderRestHook(makeCacheProvider);
    jest.useFakeTimers();
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .delay(1000)
      .reply(200, payload);
    nock('http://test.com')
      .get(`/user/`)
      .delay(2000)
      .reply(200, users);

    function MultiResourceTester() {
      try {
        const [article, user] = useResource(
          [
            CoolerArticleResource.detailShape(),
            {
              id: payload.id,
            },
          ],
          [UserResource.listShape(), {}],
        );
        return article;
      } catch (e) {
        // TODO: we're not handling suspense properly so react complains
        // When upgrading test util we should be able to fix this as we'll suspense ourselves.
        if (typeof e.then === 'function') {
          return e;
        } else {
          throw e;
        }
      }
    }
    const { rerender, result } = renderRestHook(MultiResourceTester);
    const firstPromise = result.current;
    expect(firstPromise).toBeDefined();
    expect(typeof firstPromise.then).toBe('function');
    jest.advanceTimersByTime(50);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(1000);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(2000);
    rerender();
    expect(result.current).toBe(firstPromise);

    // TODO: we're not handling suspense properly so react complains
    // When upgrading test until we should be able to fix this as we'll suspense ourselves.
    const oldError = console.error;
    console.error = () => {};
    jest.runAllTimers();
    await result.current;
    rerender();
    expect(result.current).toMatchInlineSnapshot(`
      CoolerArticleResource {
        "author": null,
        "content": "whatever",
        "id": 5,
        "tags": Array [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      }
    `);
    // eslint-disable-next-line require-atomic-updates
    console.error = oldError;
  });
  it('should NOT suspend if result already in cache and options.invalidIfStale is false', () => {
    const state = mockInitialState([
      {
        request: CoolerArticleResource.detailShape(),
        params: payload,
        result: payload,
      },
    ]);

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend even when result is stale and options.invalidIfStale is false', () => {
    const { entities, result } = normalize(
      payload,
      CoolerArticleResource.getEntitySchema(),
    );
    const fetchKey = CoolerArticleResource.detailShape().getFetchKey(payload);
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => {}}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester />
          </Suspense>{' '}
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend if result is not stale and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource.getEntitySchema(),
    );
    const fetchKey = InvalidIfStaleArticleResource.detailShape().getFetchKey(
      payload,
    );
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      meta: {
        [fetchKey]: {
          date: Infinity,
          expiresAt: Infinity,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester invalidIfStale />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should suspend if result stale in cache and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource.getEntitySchema(),
    );
    const fetchKey = InvalidIfStaleArticleResource.detailShape().getFetchKey(
      payload,
    );
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => {}}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester invalidIfStale />
          </Suspense>
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    render(tree);
    expect(fbmock).toHaveBeenCalled();
  });
});
