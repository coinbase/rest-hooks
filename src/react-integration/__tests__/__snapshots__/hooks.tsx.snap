// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCache should select paginated results 1`] = `
Array [
  PaginatedArticleResource {
    "author": null,
    "content": "the best things in life com efree",
    "id": 23,
    "tags": Array [
      "one",
      "two",
    ],
    "title": "the first draft",
  },
  PaginatedArticleResource {
    "author": null,
    "content": "the best things in life com efree",
    "id": 44,
    "tags": Array [
      "hbh",
      "wew",
    ],
    "title": "the second book",
  },
  PaginatedArticleResource {
    "author": null,
    "content": "the best things in life com efree",
    "id": 2,
    "tags": Array [
      "free",
      "honey",
    ],
    "title": "the third novel",
  },
  PaginatedArticleResource {
    "author": null,
    "content": "the best things in life com efree",
    "id": 643,
    "tags": Array [],
    "title": "a long time ago",
  },
]
`;

exports[`useFetcher should dispatch an action that fetches a create 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/rpc",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": false,
    "url": "POST http://test.com/article-cooler/",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetcher should dispatch an action that fetches a full update 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/rpc",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": false,
    "url": "PUT http://test.com/article-cooler/1",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetcher should dispatch an action that fetches a partial update 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/rpc",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": false,
    "url": "PATCH http://test.com/article-cooler/1",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should dispatch an action that fetches 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should dispatch fetch when sent multiple arguments 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should dispatch fetch when sent multiple arguments 2`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": Array [
      EntitySchema {
        "_getId": [Function],
        "_idAttribute": [Function],
        "_key": "http://test.com/user/",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "denormalize": [Function],
        "schema": Object {},
      },
    ],
    "throttle": true,
    "url": "GET http://test.com/user/",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch singles 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with fetch shape defined dataExpiryLength 1`] = `
Object {
  "meta": Object {
    "options": Object {
      "dataExpiryLength": 3600000,
    },
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-static/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-static/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with fetch shape defined errorExpiryLength 1`] = `
Object {
  "meta": Object {
    "options": Object {
      "dataExpiryLength": Infinity,
      "errorExpiryLength": Infinity,
    },
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-static/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-static/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with resource defined dataExpiryLength 1`] = `
Object {
  "meta": Object {
    "options": Object {
      "dataExpiryLength": Infinity,
    },
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-static/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-static/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;
